File: claude/research/lcg_lll.c
/* lcg_lll — lattice attack on a 64-bit LCG with high-half output, run against the
 * real archive. The Python version proved the attack works (it recovers a synthetic
 * LCG64 at K = 13 and K = 20, better than the worst-case LLL bound predicts) but its
 * exact-fraction LLL is far too slow for a sweep. This is the same attack with the
 * basis in __int128 and the Gram-Schmidt in long double: microseconds per reduction.
 *
 * Setup. bonus = first ball drawn means bonus-1 = (u*80)>>32 with u = s>>32, so each
 * state s_d is pinned to an interval of width 2^57.7 out of 2^64 — 6.32 bits. With
 * s_{d+1} = A s_d + C (mod 2^64), differencing kills the unknown increment:
 * D_d = s_{d+1} - s_d satisfies D_{d+1} = A D_d. Centring gives e_d = A^d e_0 + b_d
 * (mod 2^64) with every |e_d| <= 2^58.7, a Hidden Number Problem that LLL solves.
 * The multiplier is guessed from the standard list and W is swept.
 *
 *   ./lcg_lll selftest        positive and negative controls on synthetic data
 *   ./lcg_lll real [K] [maxW] sweep the real archive
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>
#define MAXN 64
typedef __int128 i128;
typedef unsigned __int128 u128;
static const u128 MOD = (u128)1 << 64;

static int NDIM;
static i128 Bm[MAXN][MAXN];
static long double mu[MAXN][MAXN], Bst[MAXN];

static long double dotf(const i128*a, const i128*b){
  long double s=0; for(int i=0;i<NDIM;i++) s += (long double)a[i]*(long double)b[i]; return s; }

static void gso(void){
  static long double v[MAXN][MAXN];
  for(int i=0;i<NDIM;i++){
    for(int k=0;k<NDIM;k++) v[i][k]=(long double)Bm[i][k];
    for(int j=0;j<i;j++){
      long double d=Bst[j]; long double num=0;
      for(int k=0;k<NDIM;k++) num += (long double)Bm[i][k]*v[j][k];
      mu[i][j] = d>0 ? num/d : 0.0L;
      for(int k=0;k<NDIM;k++) v[i][k] -= mu[i][j]*v[j][k];
    }
    long double s=0; for(int k=0;k<NDIM;k++) s += v[i][k]*v[i][k];
    Bst[i]=s;
  }
}
static void lll(void){
  gso();
  int k=1, guard=0;
  while(k<NDIM && guard++ < 200000){
    for(int j=k-1;j>=0;j--){
      long double q=mu[k][j];
      long double r = (q>=0)? floorl(q+0.5L) : -floorl(-q+0.5L);
      if(r!=0){ i128 ri=(i128)r;
        for(int t=0;t<NDIM;t++) Bm[k][t] -= ri*Bm[j][t];
        gso(); }
    }
    if(Bst[k] >= (0.99L - mu[k][k-1]*mu[k][k-1])*Bst[k-1]) k++;
    else { for(int t=0;t<NDIM;t++){ i128 tmp=Bm[k][t]; Bm[k][t]=Bm[k-1][t]; Bm[k-1][t]=tmp; }
           gso(); k = (k-1>1)?k-1:1; }
  }
}
static uint64_t powmod(uint64_t a,int e){ uint64_t r=1,b=a; while(e){ if(e&1)r*=b; b*=b; e>>=1; } return r; }

/* returns 1 if a vector consistent with every |e_d| <= bound is found */
static int hnp(uint64_t A, const uint64_t*cen, int K, u128 bound){
  int n=K+1; NDIM=n;
  static uint64_t Dc[MAXN], beta[MAXN];
  for(int d=0; d<K; d++) Dc[d] = cen[d+1]-cen[d];            /* mod 2^64 by wraparound */
  for(int d=1; d<K; d++) beta[d-1] = powmod(A,d)*Dc[0] - Dc[d];
  memset(Bm,0,sizeof Bm);
  for(int i=0;i<K-1;i++) Bm[i][i] = (i128)MOD;
  for(int d=1; d<K; d++) Bm[K-1][d-1] = (i128)powmod(A,d);
  Bm[K-1][K-1] = 1;
  for(int d=0; d<K-1; d++) Bm[K][d] = (i128)beta[d];
  Bm[K][K] = (i128)bound;
  lll();
  for(int i=0;i<n;i++){
    i128 last=Bm[i][K];
    if(last!=(i128)bound && last!=-(i128)bound) continue;
    int sgn = (last<0)?1:-1; int ok=1;
    for(int d=0; d<K-1 && ok; d++){ i128 e=sgn*Bm[i][d]; if(e<0)e=-e; if((u128)e>bound) ok=0; }
    i128 e0=sgn*Bm[i][K-1]; if(e0<0)e0=-e0;
    if(ok && (u128)e0<=bound) return 1;
  }
  return 0;
}
static void interval(uint32_t j,uint32_t k,uint64_t*lo,uint64_t*hi){
  u128 a=((u128)j<<32), b=((u128)(j+1)<<32);
  uint64_t l=(uint64_t)(a/k) + ((a%k)?1:0), h=(uint64_t)(b/k) + ((b%k)?1:0);
  *lo=(uint64_t)l<<32; *hi=((uint64_t)h<<32);
}
static uint64_t centre_from_bonus(int bonus){
  uint64_t lo,hi; interval((uint32_t)(bonus-1),80,&lo,&hi); return lo + ((hi-lo)>>1);
}
static const uint64_t MULT[]={6364136223846793005ULL,2862933555777941757ULL,3202034522624059733ULL,
  3935559000370003845ULL,1181783497276652981ULL,0xda942042e4dd58b5ULL,25214903917ULL,
  1103515245ULL,214013ULL,1664525ULL,6364136223846793005ULL,0x27BB2EE687B0B0FDULL};
static const char*MNAME[]={"MMIX / PCG","L'Ecuyer a","L'Ecuyer b","L'Ecuyer c","ranqd1-64",
  "Lehmer64","drand48","glibc LCG","MSVC","Numerical Recipes","PCG-XSL-RR","Knuth 64"};
#define NMULT 12

static uint32_t N,*IDS,*TS; static uint8_t *NUMS,*BOOST,*BONUS;
static void loadbin(void){
  FILE*f=fopen("draws.bin","rb"); if(!f||fread(&N,4,1,f)!=1){perror("draws.bin");exit(1);}
  IDS=malloc(4*N);TS=malloc(4*N); uint64_t*LO=malloc(8*N),*HI=malloc(8*N);
  NUMS=malloc(20*N);BOOST=malloc(N);BONUS=malloc(N);
  if(fread(IDS,4,N,f)!=N||fread(TS,4,N,f)!=N||fread(LO,8,N,f)!=N||fread(HI,8,N,f)!=N||
     fread(NUMS,20,N,f)!=N||fread(BOOST,1,N,f)!=N||fread(BONUS,1,N,f)!=N){fprintf(stderr,"short read\n");exit(1);}
  fclose(f);
}
int main(int argc,char**argv){
  const char*mode = argc>1?argv[1]:"selftest";
  uint64_t lo,hi; interval(0,80,&lo,&hi); u128 bound=(u128)(hi-lo);
  if(!strcmp(mode,"selftest")){
    uint64_t a=6364136223846793005ULL, c=1442695040888963407ULL; int W=21;
    printf("selftest: synthetic LCG64 (MMIX), W=%d, output s>>32, bonus = first ball\n",W);
    printf("  positive AND negative controls must both pass before any real run\n\n");
    for(int K=12;K<=24;K+=4){
      static uint64_t cen[MAXN]; uint64_t s=0xC0FFEE1234567890ULL;
      for(int d=0;d<=K;d++){ uint32_t u=(uint32_t)(s>>32); uint32_t j=(uint32_t)(((uint64_t)u*80)>>32);
        uint64_t l,h; interval(j,80,&l,&h); cen[d]=l+((h-l)>>1);
        for(int t=0;t<W;t++) s=a*s+c; }
      int good=hnp(powmod(a,W),cen,K,bound);
      int badW=hnp(powmod(a,W+1),cen,K,bound);
      int badA=hnp(powmod(2862933555777941757ULL,W),cen,K,bound);
      static uint64_t rnd[MAXN]; uint64_t z=1;
      for(int d=0;d<=K;d++){ z=z*6364136223846793005ULL+1442695040888963407ULL; rnd[d]=z; }
      int badD=hnp(powmod(a,W),rnd,K,bound);
      printf("  K=%2d  correct a,W: %-10s | wrong W: %-9s | wrong a: %-9s | random data: %s\n",
        K, good?"RECOVERED":"missed", badW?"false hit":"rejected",
        badA?"false hit":"rejected", badD?"false hit":"rejected");
    }
    return 0;
  }
  loadbin();
  int K=argc>2?atoi(argv[2]):20, maxW=argc>3?atoi(argv[3]):48;
  int starts[]={0,5000,20000,50000,70000-40};
  printf("real archive: %u draws, K=%d constraints, W swept 1..%d, %d start offsets\n",N,K,maxW,5);
  printf("  bonus = first ball drawn, LCG64 output s>>32, increment differenced away\n\n");
  int total=0;
  for(int m=0;m<NMULT;m++){
    int found=0;
    for(int W=1;W<=maxW;W++){
      uint64_t A=powmod(MULT[m],W);
      for(int si=0;si<5;si++){ int st=starts[si];
        static uint64_t cen[MAXN];
        for(int d=0;d<=K;d++) cen[d]=centre_from_bonus(BONUS[st+d]);
        if(hnp(A,cen,K,bound)){ printf("  !!! HIT %s W=%d start=%d\n",MNAME[m],W,st); found++; }
      }
    }
    printf("  %-20s a=%-22llu %s\n",MNAME[m],(unsigned long long)MULT[m],
           found?"SEE HITS ABOVE":"no fit at any W");
    total+=found;
  }
  printf("\n  total hits: %d  ->  %s\n",total,
     total?"INVESTIGATE":"no 64-bit LCG with a standard multiplier fits the archive");
  return 0;
}

File: claude/research/lcg_lll.py
"""Feasibility of a lattice attack on a 64-bit LCG whose output is the HIGH half.

This is the one classical family the rest of the work does not reach: seedhunt's 2^32
sweep covers every LCG modulo 2^32 and every 64-bit LCG reachable from a 32-bit seed;
the F2-linear attacks do not apply to a congruential update; modlcg only sees a
congruential generator whose output carries the LOW bits. What is left is an LCG
modulo 2^64 with an arbitrary 64-bit state and output s >> 32.

The bonus channel pins each state: bonus-1 = (u*80)>>32 with u = s >> 32 puts s in an
interval of width 2^57.7 out of 2^64, i.e. 6.3 bits per draw. Differencing removes the
unknown increment, leaving a Hidden Number Problem, and LLL is the standard tool.

It does not work here, and the script measures exactly why rather than asserting it.
Centre every quantity: with e_d = D_d - Dc_d bounded by 2^58.7 and e_d = A^d e_0 + b_d
(mod 2^64), the target vector has norm about sqrt(K+1)*2^58.7 while the Gaussian
heuristic for the lattice sits at ((2^64)^(K-1) * 2^58.7)^(1/(K+1)). The target only
becomes the shortest vector past K ~ 30, and by then LLL's approximation factor
(~2^(n/4), so 2^7 or worse) swamps the 2^0.6 margin. Classical truncated-LCG attacks
assume half the bits of each output are visible; 6.3 bits out of 64 is far below that.

Conclusion recorded honestly: this family is NOT excluded. It is the one gap left, and
capturing the draw order closes it — 20 ordered indices per draw is 126 bits, well
past what the lattice needs.

Usage: python3 lcg_lll.py selftest   |   python3 lcg_lll.py margins
"""
import sys, math
from fractions import Fraction

M = 1 << 64

def lll(B, delta=Fraction(99, 100)):
    """Integer LLL with exact rational Gram-Schmidt, GSO refreshed only on swaps."""
    B = [row[:] for row in B]
    n = len(B)
    def dot(u, v): return sum(x*y for x, y in zip(u, v))
    def gso():
        Bs = []; mu = [[Fraction(0)]*n for _ in range(n)]
        for i in range(n):
            v = [Fraction(x) for x in B[i]]
            for j in range(i):
                d = dot(Bs[j], Bs[j])
                mu[i][j] = Fraction(dot([Fraction(x) for x in B[i]], Bs[j]), d) if d else Fraction(0)
                v = [a - mu[i][j]*b for a, b in zip(v, Bs[j])]
            Bs.append(v)
        return Bs, mu
    Bs, mu = gso()
    k = 1; guard = 0
    while k < n and guard < 40000:
        guard += 1
        changed = False
        for j in range(k-1, -1, -1):
            q = mu[k][j]
            r = int(q + Fraction(1, 2)) if q >= 0 else -int(-q + Fraction(1, 2))
            if r:
                B[k] = [a - r*b for a, b in zip(B[k], B[j])]
                changed = True
        if changed:
            Bs, mu = gso()
        if dot(Bs[k], Bs[k]) >= (delta - mu[k][k-1]**2) * dot(Bs[k-1], Bs[k-1]):
            k += 1
        else:
            B[k], B[k-1] = B[k-1], B[k]
            Bs, mu = gso()
            k = max(k-1, 1)
    return B

def hnp(A, cen, bound):
    """Centred HNP: find e_0 with |e_d| <= bound where e_d = A^d e_0 + b_d (mod M)."""
    K = len(cen) - 1
    Dc = [(cen[d+1] - cen[d]) % M for d in range(K)]
    beta = [(pow(A, d, M)*Dc[0] - Dc[d]) % M for d in range(1, K)]
    n = K + 1
    rows = []
    for i in range(K-1):
        rows.append([M if j == i else 0 for j in range(K-1)] + [0, 0])
    rows.append([pow(A, d, M) for d in range(1, K)] + [1, 0])
    rows.append(beta + [0, bound])
    red = lll(rows)
    for v in red:
        if abs(v[-1]) != bound:
            continue
        sgn = 1 if v[-1] < 0 else -1
        e = [sgn*x for x in v[:K-1]]
        e0 = sgn*v[-2]
        if abs(e0) <= bound and all(abs(x) <= bound for x in e):
            return e0
    return None

def interval(j, k=80):
    lo = (j << 32)//k + (1 if ((j << 32) % k) else 0)
    hi = ((j+1) << 32)//k + (1 if (((j+1) << 32) % k) else 0)
    return lo << 32, hi << 32

def selftest(Ks=(13, 20), a=6364136223846793005, c=1442695040888963407, W=21):
    print("selftest: synthetic LCG64 (MMIX), W=%d, output s>>32, bonus = first ball" % W)
    print("  a positive result is required before any negative result is reported\n")
    A = pow(a, W, M)
    for K in Ks:
        s = 0xC0FFEE1234567890
        cen = []
        for d in range(K+1):
            lo, hi = interval((( (s >> 32)*80) >> 32))
            assert lo <= s < hi
            cen.append((lo+hi)//2)
            for _ in range(W):
                s = (a*s + c) % M
        bound = 2*((interval(0)[1]-interval(0)[0])//2)
        got = hnp(A, cen, bound)
        # feasibility numbers
        n = K+1
        logdet = 64*(K-1) + math.log2(bound)
        gauss = math.log2(math.sqrt(n/(2*math.pi*math.e))) + logdet/n
        tgt = math.log2(math.sqrt(n)) + math.log2(bound)
        print("  K=%2d : target norm 2^%.1f   Gaussian heuristic 2^%.1f   margin 2^%+.1f "
              "  LLL factor 2^%.1f   -> %s"
              % (K, tgt, gauss, gauss-tgt, n/4.0, "RECOVERED" if got is not None else "missed"))
    print("\n  The margin never gets near LLL's approximation factor, so the attack cannot")
    print("  work at 6.3 known bits per draw. This family stays open — see the write-up.")

def margins():
    bound = 2*((interval(0)[1]-interval(0)[0])//2)
    print("bits known per draw from the bonus channel: %.2f" % (64 - math.log2(bound)))
    print(" K   dim   target    Gaussian   margin    LLL factor needed")
    for K in (13, 20, 30, 40, 60, 80, 120):
        n = K+1
        logdet = 64*(K-1) + math.log2(bound)
        gauss = math.log2(math.sqrt(n/(2*math.pi*math.e))) + logdet/n
        tgt = math.log2(math.sqrt(n)) + math.log2(bound)
        print("%3d  %4d  2^%.1f   2^%.1f   2^%+.1f    2^%.1f" % (K, n, tgt, gauss, gauss-tgt, n/4.0))
    print("\nLLL delivers roughly 2^(n/4); the margin has to exceed that. It never does.")

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "margins": margins()
    else: selftest()
